core/src/main/java/org/apache/calcite/sql/fun/SqlCastFunction.java

```diff
@@ -122,8 +128,50 @@ static SqlReturnTypeInference returnTypeInference(boolean safe) {
   /** Derives the type of "CAST(expression AS targetType)". */
   public static RelDataType deriveType(RelDataTypeFactory typeFactory,
       RelDataType expressionType, RelDataType targetType, boolean safe) {
-    return typeFactory.createTypeWithNullability(targetType,
-        expressionType.isNullable() || safe);
+    return createTypeWithNullabilityFromExpr(typeFactory, expressionType, targetType, safe);
+  }
+
+  private static RelDataType createTypeWithNullabilityFromExpr(RelDataTypeFactory typeFactory,
+      RelDataType expressionType, RelDataType targetType, boolean safe) {
+    boolean isNullable = expressionType.isNullable() || safe;
+
+    if (isCollection(expressionType)) {
+      RelDataType expressionElementType = expressionType.getComponentType();
+      RelDataType targetElementType = targetType.getComponentType();
+      RelDataType newElementType =
+          createTypeWithNullabilityFromExpr(typeFactory, expressionElementType, targetElementType, false);
+      if (isArray(expressionType)) {
+        return SqlTypeUtil.createArrayType(typeFactory, newElementType, isNullable);
+      } else {
+        return SqlTypeUtil.createMultisetType(typeFactory, newElementType, isNullable);
```

Comment:
Maybe we can use `isArray(expressionType) ?  xxx: yyy` to be more concise.

Reply:
fixed @chunweilei